graphium/trainer/predictor.py

```diff
@@ -449,35 +447,47 @@ def flag_step(self, batch: Dict[str, Tensor], step_name: str, to_cpu: bool) -> D
-        if (batch_idx + 1) % self.metrics_every_n_steps != 0:
-            return
 
+        # Get the throughput of the batch
         train_batch_time = time.time() - self.train_batch_start_time
         num_graphs = self.get_num_graphs(batch["features"])
         tput = num_graphs / train_batch_time
-        if self.save_preds_and_targets:
-            # this code is likely repeated for validation and testing, this should be moved to a function
-            self.task_epoch_summary.update_predictor_state(
-                step_name="train",
-                targets=outputs["targets"],
-                predictions=outputs["preds"],
-                loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
-                task_losses=outputs["task_losses"],
-                n_epochs=self.current_epoch,
-            )
-            metrics_logs = self.task_epoch_summary.get_metrics_logs()  # Dict[task, metric_logs]
-            metrics_logs["_global"]["grad_norm"] = self.get_gradient_norm()
-            outputs.update(metrics_logs)  # Dict[task, metric_logs]. Concatenate them?
 
+        # Get the metrics that are logged at every step (loss, grad_norm, batch_time, batch_tput)
         concatenated_metrics_logs = {}  # self.task_epoch_summary.concatenate_metrics_logs(metrics_logs)
         concatenated_metrics_logs["train/loss"] = outputs["loss"]
         outputs["grad_norm"] = self.get_gradient_norm()
         concatenated_metrics_logs["train/grad_norm"] = outputs["grad_norm"]
         concatenated_metrics_logs["train/batch_time"] = train_batch_time
         concatenated_metrics_logs["train/batch_tput"] = tput
+
+        # If logging is skipped for this step, then log the important metrics anyway and return
+        if self.is_log_train_metrics:
+            if self.logger is not None:
+                self.logger.log_metrics(
+                    concatenated_metrics_logs, step=self.global_step
+                )  # This is a pytorch lightning function call
+            return
+
+        # this code is likely repeated for validation and testing, this should be moved to a function
+        self.task_epoch_summary.update_predictor_state(
+            step_name="train",
+            targets=outputs["targets"],
+            predictions=outputs["preds"],
+            loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
+            task_losses=outputs["task_losses"],
+            n_epochs=self.current_epoch,
+        )
+        metrics_logs = self.task_epoch_summary.get_metrics_logs()  # Dict[task, metric_logs]
+        metrics_logs["_global"]["grad_norm"] = self.get_gradient_norm()
+        outputs.update(metrics_logs)  # Dict[task, metric_logs]. Concatenate them?
+
         # report the training loss for each individual tasks
         for task in self.tasks:
             concatenated_metrics_logs[f"train/loss/{task}"] = outputs["task_losses"][task]
```

Comment:
I tested the debug config with `metrics_every_n_train_steps: 30` but somehow the individual task losses were not reported (see [here](https://wandb.ai/multitask-gnn/neurips2023_large_data_gcn_debug/runs/hss92bbw?workspace=user-zhiyi-li1230)), just wondering if there's anything missing in my setting?

Reply:
See my comment below, this is a draft I have made and could have mistakes. Feel free to edit this branch before merging

Reply:
I think I fixed that in the latest commit, I moved that part before the `if self.skip_log_train_metrics`

Reply:
I also moved the `concatenated_metrics_logs[key] = concatenated_metrics_logs[key].mean()` part up so that we get averaged loss for the individual losses.